### DISCUSSION TITLE
Calculating reduced density matrix of each site

### DISCUSSION THREAD

**Discussion by Shovan-Physics** [block-hczhai/block2-preview#123] [Q&A] [answered]
Is there an efficient way to get the reduced density matrix of each site? For example, if I want the distribution over the local basis states (in the Bose-Hubbard model), currently I am defining projectors onto each of the basis states, forming MPO with each projector, and calculating their expectation values, which is rather clumsy and repetitive.

**Comment by hczhai** [comment]
This is solved in https://github.com/block-hczhai/block2-preview/discussions/122 using custom operators.

**Reply by Shovan-Physics** [reply]
Thanks for the update. Although it makes the calculation faster, it doesn't quite solve this problem: 

Suppose I have 6 local basis states (as in the Bose-Hubbard model with `NB_MAX = 5`) and I want to get the 6 x 6 reduced density matrix of each site (e.g., to compare with a Gutzwiller ansatz). Currently, I need to define 20 on-site operators and run `get_npdm` for each of them to get all the independent matrix elements.

However, if we could align the MPS center at a site and contract all other site and bond indices we could get the density matrix $\rho^{(j)}_{\sigma, \sigma^{\prime}}$ at one go. Is this possible to do? I know there is a command `align_mps_center` but I don't know how to get/contract the indices.

<img width="606" alt="reduced_dm_mps" src="https://github.com/user-attachments/assets/19e4c2d7-10bc-46a8-8fc0-236cf3d205de">

**Reply by hczhai** [reply, marked as answer]
You can simply do

```python3
dms = driver.get_npdm(mps, npdm_expr=['N', 'M', 'L', 'K', 'J', 'I'], fermionic_ops='')
```

which will generate results of

```python3
dms = [
    *driver.get_npdm(mps, npdm_expr='N', fermionic_ops=''),
    *driver.get_npdm(mps, npdm_expr='M', fermionic_ops=''),
    ...
    *driver.get_npdm(mps, npdm_expr='I', fermionic_ops=''),
]
```

at one time.

**Reply by Shovan-Physics** [reply]
That helps, thanks. Do we need to use a single letter for each operator? I tried using `'P0','P1','P2'..` as well as `'0','1','2'..` but got error. Wondering in case I run out of the alphabet.

**Reply by hczhai** [reply]
Numbers have special meaning in SU(2) strings so they cannot be used for operators.

> Wondering in case I run out of the alphabet.

You can define a subset of your operators, call ``driver.get_npdm``, then redefine ``site_ops`` for the remaining operators, and call ``driver.get_npdm``, etc.

**Reply by Shovan-Physics** [reply]
OK thanks.
